nosnoc/problem.py

```diff
@@ -186,14 +186,36 @@ def create_complementarity(self, x: List[ca.SX], y: ca.SX, sigma: ca.SX, tau: ca
 
 
 class FiniteElementBase(NosnocFormulationObject):
+    def X_fe(self) -> List[ca.SX]:
+        
+        """returns list of all x values in finite element"""
+        opts = NosnocOpts()
+        if opts.irk_representation == IrkRepresentation.INTEGRAL:
+            X_fe = [self.w[ind] for ind in self.ind_x]
+        elif opts.irk_representation == IrkRepresentation.DIFFERENTIAL:
+            X_fe = []
+            for j in range(opts.n_s):
+                x_temp = self.prev_fe.w[self.prev_fe.ind_x[-1]]
+                for r in range(opts.n_s):
+                    x_temp += self.h * opts.A_irk[j, r] * self.w[self.ind_v[r]]
+                X_fe.append(x_temp)
+            X_fe.append(self.w[self.ind_x[-1]])
+        elif opts.irk_representation == IrkRepresentation.DIFFERENTIAL_LIFT_X:
+            X_fe = [self.w[ind] for ind in self.ind_x]
+
+        return X_fe
 
     def Lambda(self, stage=slice(None), sys=slice(None)):
         return ca.vertcat(self.w[flatten(self.ind_lam[stage][sys])],
                           self.w[flatten(self.ind_lambda_n[stage][sys])],
                           self.w[flatten(self.ind_lambda_p[stage][sys])])
     
-    def C_pds(self, stage=slice(None), sys=slice(None)):
-        return ca.vertcat(self.w[flatten(self.ind_c_pds[stage][sys])])
+    def C_pds(self, stage=slice(None), sys=slice(None)) -> ca.SX:
+    
+        x = self.X_fe()[stage][sys]  
+        c_pds = self.model.c_pds_fun(x)
+        c_pds = ca.if_else(c_pds < 0, 0, c_pds)
```

Comment:
This isn't adding a lower bound, it just means that the function $c(x)$ itself only takes positive values, which is not what we want. You should add a constraint at each stage using `add_constraint` in the `forward_simulation` method.